// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/data-sources/parsing.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParsingStategy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const default_parsing_prompt_1 = require("./default-parsing-prompt");
/**
 * Enum representing the types of parsing strategies available for Amazon Bedrock Knowledge Bases.
 */
var ParsingStategyType;
(function (ParsingStategyType) {
    /**
     * Uses a Bedrock Foundation Model for advanced parsing of non-textual information from documents.
     */
    ParsingStategyType["FOUNDATION_MODEL"] = "BEDROCK_FOUNDATION_MODEL";
})(ParsingStategyType || (ParsingStategyType = {}));
/**
 * Represents an advanced parsing strategy configuration for Knowledge Base ingestion.
 * @see https://docs.aws.amazon.com/bedrock/latest/userguide/kb-chunking-parsing.html#kb-advanced-parsing
 */
class ParsingStategy {
    // ------------------------------------------------------
    // FM Parsing Strategy
    // ------------------------------------------------------
    /**
     * Creates a Foundation Model-based parsing strategy for extracting non-textual information
     * from documents such as tables and charts.
     * - Additional costs apply when using advanced parsing due to foundation model usage.
     * - There are limits on file types (PDF) and total data that can be parsed using advanced parsing.
     * @see https://docs.aws.amazon.com/bedrock/latest/userguide/knowledge-base-ds.html#kb-ds-supported-doc-formats-limits
     */
    static foundationModel(props) {
        class FoundationModelTransformation extends ParsingStategy {
            constructor() {
                super(...arguments);
                /** The CloudFormation property representation of this configuration */
                this.configuration = {
                    bedrockFoundationModelConfiguration: {
                        modelArn: props.parsingModel.invokableArn,
                        parsingPrompt: {
                            parsingPromptText: props.parsingPrompt ?? default_parsing_prompt_1.DEFAULT_PARSING_PROMPT,
                        },
                    },
                    parsingStrategy: ParsingStategyType.FOUNDATION_MODEL,
                };
            }
            generatePolicyStatements() {
                return [
                    new aws_iam_1.PolicyStatement({
                        actions: ['bedrock:InvokeModel'],
                        resources: [props.parsingModel.invokableArn],
                    }),
                ];
            }
        }
        return new FoundationModelTransformation();
    }
}
exports.ParsingStategy = ParsingStategy;
_a = JSII_RTTI_SYMBOL_1;
ParsingStategy[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.ParsingStategy", version: "0.1.285" };
