// @cdklabs/generative-ai-cdk-constructs/lib/common/helpers/dynamodb-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createKendraSyncRunTable = void 0;
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const cdk = require("aws-cdk-lib");
const aws_cdk_lib_1 = require("aws-cdk-lib");
function createKendraSyncRunTable(cdkStack) {
    return new cdk.aws_dynamodb.Table(cdkStack, 'SyncRunTable', {
        tableName: 'SyncRunTable',
        partitionKey: {
            name: 'Id',
            type: cdk.aws_dynamodb.AttributeType.STRING,
        },
        sortKey: {
            name: 'CreatedOn',
            type: cdk.aws_dynamodb.AttributeType.STRING,
        },
        encryption: cdk.aws_dynamodb.TableEncryption.AWS_MANAGED,
        billingMode: cdk.aws_dynamodb.BillingMode.PAY_PER_REQUEST,
        removalPolicy: aws_cdk_lib_1.RemovalPolicy.DESTROY,
    });
}
exports.createKendraSyncRunTable = createKendraSyncRunTable;
