// @cdklabs/generative-ai-cdk-constructs/lib/common/helpers/python-lambda-layer-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Layer = void 0;
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const child_process_1 = require("child_process");
const cdk = require("aws-cdk-lib");
const lambda = require("aws-cdk-lib/aws-lambda");
const s3assets = require("aws-cdk-lib/aws-s3-assets");
const constructs_1 = require("constructs");
class Layer extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const { runtime, architecture, path, additionalPackages, autoUpgrade, local } = props;
        const args = local ? [] : ['-t /asset-output/python'];
        if (additionalPackages) {
            args.push(...additionalPackages);
        }
        if (autoUpgrade) {
            args.push('--upgrade');
        }
        const layerAsset = new s3assets.Asset(this, 'LayerAsset', {
            path,
            bundling: local ? {
                // If local is true use the host to install the requirements
                image: runtime.bundlingImage,
                local: {
                    tryBundle(outputDir) {
                        (0, child_process_1.execSync)(`${local} -m venv venv`);
                        (0, child_process_1.execSync)('source venv/bin/activate');
                        (0, child_process_1.execSync)(`pip install -r ${path}/requirements.txt -t ${outputDir}/python ${args.join(' ')}`);
                        return true;
                    },
                },
            } : {
                // Default: Docker is used to install the requirements
                image: runtime.bundlingImage,
                platform: architecture.dockerPlatform,
                command: [
                    'bash',
                    '-c',
                    `pip install -r requirements.txt ${args.join(' ')}`,
                ],
                outputType: cdk.BundlingOutput.AUTO_DISCOVER,
                securityOpt: 'no-new-privileges:true',
                network: 'host',
            },
        });
        const layer = new lambda.LayerVersion(this, 'Layer', {
            code: lambda.Code.fromBucket(layerAsset.bucket, layerAsset.s3ObjectKey),
            compatibleRuntimes: [runtime],
            compatibleArchitectures: [architecture],
            ...props,
        });
        this.layer = layer;
    }
}
exports.Layer = Layer;
