// @cdklabs/generative-ai-cdk-constructs/lib/common/helpers/redis-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckRedisClusterProps = exports.setInboundRules = exports.getRedisSecurityGroup = exports.buildRedisCluster = void 0;
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const ec2 = require("aws-cdk-lib/aws-ec2");
const elasticache = require("aws-cdk-lib/aws-elasticache");
function buildRedisCluster(scope, props) {
    const cacheNodeType = props.cfnCacheClusterProps?.cacheNodeType || 'cache.r6g.xlarge';
    const engine = props.cfnCacheClusterProps?.engine || 'redis';
    const numCacheNodes = props.cfnCacheClusterProps?.numCacheNodes || 1;
    let redisCulster = new elasticache.CfnCacheCluster(scope, 'redisCluster', {
        cacheNodeType: cacheNodeType,
        engine: engine,
        numCacheNodes: numCacheNodes,
        cacheSubnetGroupName: getRedisSubnetGroup(scope, props).ref,
        vpcSecurityGroupIds: [props.redisSecurityGroup.securityGroupId],
        port: props.redisPort,
    });
    return redisCulster;
}
exports.buildRedisCluster = buildRedisCluster;
// get redis subnet group from existing vpc
function getRedisSubnetGroup(scope, props) {
    let redisSubnetGroup = new elasticache.CfnSubnetGroup(scope, 'redisSubnetGroup', {
        description: 'Redis subnet group',
        subnetIds: props.subnetIds,
    });
    return redisSubnetGroup;
}
// get redis security group from existing vpc
function getRedisSecurityGroup(scope, props) {
    let redisSecurityGroup = new ec2.SecurityGroup(scope, 'redisSecurityGroup', {
        vpc: props.existingVpc,
        allowAllOutbound: true,
        description: 'security group for elasticache',
    });
    return redisSecurityGroup;
}
exports.getRedisSecurityGroup = getRedisSecurityGroup;
function setInboundRules(redisSecurityGroup, sourceSecuritygroup, redisPort) {
    redisSecurityGroup.connections.allowFrom(sourceSecuritygroup, ec2.Port.tcp(redisPort));
}
exports.setInboundRules = setInboundRules;
function CheckRedisClusterProps(propsObject) {
    let errorMessages = '';
    let errorFound = false;
    if (propsObject.existingRedisCulster && propsObject.cfnCacheClusterProps) {
        errorMessages += 'Error - Either provide existingRedisCulster or cfnCacheClusterProps, but not both.\n';
        errorFound = true;
    }
    if (errorFound) {
        throw new Error(errorMessages);
    }
}
exports.CheckRedisClusterProps = CheckRedisClusterProps;
