// @cdklabs/generative-ai-cdk-constructs/lib/common/base-class/construct-name-enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructName = void 0;
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
var ConstructName;
(function (ConstructName) {
    ConstructName["AWSRAGAPPSYNCSTEPFNOPENSEARCH"] = "C1";
    ConstructName["AWSQAAPPSYNCOPENSEARCH"] = "C2";
    ConstructName["AWSSUMMARIZATIONAPPSYNCSTEPFN"] = "C3";
    ConstructName["AWSMODELDEPLOYMENTSAGEMAKER"] = "C4";
    ConstructName["CUSTOMSAGEMAKERENDPOINT"] = "C5";
    ConstructName["HUGGINGFACESAGEMAKERENDPOINT"] = "C6";
    ConstructName["JUMPSTARTSAGEMAKERENDPOINT"] = "C7";
    ConstructName["AWSCONTENTGENAPPSYNCLAMBDA"] = "C8";
    ConstructName["AWSTEXTTOSQL"] = "C11";
    ConstructName["AWSLLAMAINDEXDATALOADER"] = "C12";
})(ConstructName || (exports.ConstructName = ConstructName = {}));
