// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/opensearchserverless/analysis-plugins.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenFilterType = exports.TokenizerType = exports.CharacterFilterType = void 0;
var CharacterFilterType;
(function (CharacterFilterType) {
    CharacterFilterType["ICU_NORMALIZER"] = "icu_normalizer";
})(CharacterFilterType || (exports.CharacterFilterType = CharacterFilterType = {}));
// Currently we only support Kuromoji and ICU tokenizers.
// Also see the following link for more information regarding supported plugins:
// https://docs.aws.amazon.com/opensearch-service/latest/developerguide/serverless-genref.html#serverless-plugins
var TokenizerType;
(function (TokenizerType) {
    TokenizerType["KUROMOJI_TOKENIZER"] = "kuromoji_tokenizer";
    TokenizerType["ICU_TOKENIZER"] = "icu_tokenizer";
})(TokenizerType || (exports.TokenizerType = TokenizerType = {}));
var TokenFilterType;
(function (TokenFilterType) {
    TokenFilterType["KUROMOJI_BASEFORM"] = "kuromoji_baseform";
    TokenFilterType["KUROMOJI_PART_OF_SPEECH"] = "kuromoji_part_of_speech";
    TokenFilterType["KUROMOJI_STEMMER"] = "kuromoji_stemmer";
    TokenFilterType["CJK_WIDTH"] = "cjk_width";
    TokenFilterType["JA_STOP"] = "ja_stop";
    TokenFilterType["LOWERCASE"] = "lowercase";
    TokenFilterType["ICU_FOLDING"] = "icu_folding";
})(TokenFilterType || (exports.TokenFilterType = TokenFilterType = {}));
