// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/agent-alias.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentAlias = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const aws_cdk_lib_1 = require("aws-cdk-lib");
const constructs_1 = require("constructs");
class AgentAlias extends constructs_1.Construct {
    // ------------------------------------------------------
    // Imports
    // ------------------------------------------------------
    /**
     * Brings an Agent Alias from an existing one created outside of CDK.
     */
    static fromAliasArn(scope, id, aliasArn) {
        class Import extends aws_cdk_lib_1.Resource {
            constructor() {
                super(scope, id);
                this.aliasArn = aliasArn;
                [this.agentId, this.aliasId] = this.parseArnComponents(aliasArn);
            }
            parseArnComponents(arn) {
                const resourceName = aws_cdk_lib_1.Arn.split(arn, aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_SLASH_RESOURCE_NAME).resourceName;
                const [agentId, aliasId] = resourceName.split('/');
                return [agentId, aliasId];
            }
        }
        return new Import();
    }
    constructor(scope, id, props) {
        super(scope, id);
        const alias = new aws_cdk_lib_1.aws_bedrock.CfnAgentAlias(this, 'MyCfnAgentAlias', {
            agentAliasName: props.aliasName ?? 'latest',
            agentId: props.agentId,
            tags: props.tags,
            description: props.description,
        });
        if (props.agentVersion) {
            alias.routingConfiguration = [{
                    agentVersion: props.agentVersion,
                }];
        }
        this.agentId = props.agentId;
        this.aliasId = alias.attrAgentAliasId;
        this.aliasArn = alias.attrAgentAliasArn;
        this.aliasName = props.aliasName ?? 'latest';
    }
}
exports.AgentAlias = AgentAlias;
_a = JSII_RTTI_SYMBOL_1;
AgentAlias[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.AgentAlias", version: "0.1.285" };
