// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/api-schema.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3ApiSchema = exports.InlineApiSchema = exports.ApiSchema = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const fs = require("fs");
/**
 * Bedrock Agents Action Group API Schema definition.
 */
class ApiSchema {
    /**
     * Inline code for API Schema
     * @return `InlineApiSchema` with inline schema
     * @param schema The actual Open API schema
     */
    static fromInline(schema) {
        return new InlineApiSchema(schema);
    }
    /**
     * Loads the API Schema from a local disk path.
     * @return `InlineApiSchema` with the contents of `path`
     * @param path Path to the Open API schema file in yaml or JSON
     */
    static fromAsset(path) {
        return new InlineApiSchema(fs.readFileSync(path, 'utf8'));
    }
    /**
     * API Schema as an S3 object.
     * @return `S3ApiSchema` with the S3 bucket and key.
     * @param bucket The S3 bucket
     * @param key The object key
     */
    static fromBucket(bucket, key) {
        return new S3ApiSchema(bucket, key);
    }
}
exports.ApiSchema = ApiSchema;
_a = JSII_RTTI_SYMBOL_1;
ApiSchema[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.ApiSchema", version: "0.1.285" };
/**
 * API Schema from a string value.
 */
class InlineApiSchema extends ApiSchema {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    bind(_scope) {
        return {
            payload: this.schema,
        };
    }
}
exports.InlineApiSchema = InlineApiSchema;
_b = JSII_RTTI_SYMBOL_1;
InlineApiSchema[_b] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.InlineApiSchema", version: "0.1.285" };
/**
 * API Schema in an S3 object.
 */
class S3ApiSchema extends ApiSchema {
    constructor(bucket, key) {
        super();
        this.bucket = bucket;
        this.key = key;
    }
    bind(_scope) {
        return {
            s3: {
                s3BucketName: this.bucket.bucketName,
                s3ObjectKey: this.key,
            },
        };
    }
}
exports.S3ApiSchema = S3ApiSchema;
_c = JSII_RTTI_SYMBOL_1;
S3ApiSchema[_c] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.S3ApiSchema", version: "0.1.285" };
